core/kernel/user_access.c

```diff
@@ -68,8 +90,79 @@ TEE_Result copy_to_user_private(void *uaddr, const void *kaddr, size_t len)
 
 	uaddr = memtag_strip_tag(uaddr);
 	res = check_access(flags, uaddr, len);
-	if (!res)
+	if (!res) {
+		enter_user_access();
 		memcpy(uaddr, kaddr, len);
+		exit_user_access();
+	}
+
+	return res;
+}
+
+TEE_Result clear_user(void *uaddr, size_t n)
+{
+	uint32_t flags = TEE_MEMORY_ACCESS_WRITE | TEE_MEMORY_ACCESS_ANY_OWNER;
+	TEE_Result res = TEE_SUCCESS;
+
+	uaddr = memtag_strip_tag(uaddr);
+	res = check_access(flags, uaddr, n);
+	if (res)
+		return res;
+
+	enter_user_access();
+	memset(uaddr, 0, n);
+	exit_user_access();
+
+	return TEE_SUCCESS;
+}
+
+size_t strnlen_user(const void *uaddr, size_t len)
+{
+	uint32_t flags = TEE_MEMORY_ACCESS_READ | TEE_MEMORY_ACCESS_ANY_OWNER;
+	TEE_Result res = TEE_SUCCESS;
+	size_t n = 0;
+
+	uaddr = memtag_strip_tag_const(uaddr);
+	res = check_access(flags, uaddr, len);
+	if (!res) {
+		enter_user_access();
+		n = strnlen(uaddr, len);
+		exit_user_access();
+	}
+
+	return n;
+}
+
+TEE_Result memdup_user(const void *src, size_t len, void **p)
+{
+	TEE_Result res = TEE_SUCCESS;
+
+	*p = malloc(len);
```

Comment:
all good, but it would be bettre `*p` is unaffected upon failure.
suggestion below (also applies to  `memdup_user_private()`) but feel free to discard my comment:
```c
	TEE_Result res = TEE_SUCCESS;
	void *buf = NULL;

	buf = malloc(len);
	if (!buf)
		return TEE_ERROR_OUT_OF_MEMORY;

	res = copy_from_user(buf, src, len);
	if (res)
		free(buf);
	else
		*p = buf;

	return res;
```

Reply:
updated both comment and code, thanks!